core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java

```diff
@@ -527,17 +527,46 @@ void createMeasurementEvent(
                     }
                     case NewLineProtoParser.ENTITY_TYPE_STRING:
                     case NewLineProtoParser.ENTITY_TYPE_LONG256: {
-                        Unsafe.getUnsafe().putByte(bufPos, entity.getType());
-                        bufPos += Byte.BYTES;
-                        int l = entity.getValue().length();
-                        Unsafe.getUnsafe().putInt(bufPos, l);
-                        bufPos += Integer.BYTES;
-                        long hi = bufPos + 2L * l;
-                        floatingCharSink.of(bufPos, hi);
-                        if (!Chars.utf8Decode(entity.getValue().getLo(), entity.getValue().getHi(), floatingCharSink)) {
-                            throw CairoException.instance(0).put("invalid UTF8 in value for ").put(entity.getName());
+                        final int colTypeMeta = localDetails.getColumnTypeMeta(colIndex);
+                        if (colTypeMeta == -1) {
+                            Unsafe.getUnsafe().putByte(bufPos, entity.getType());
+                            bufPos += Byte.BYTES;
+                            int l = entity.getValue().length();
+                            Unsafe.getUnsafe().putInt(bufPos, l);
+                            bufPos += Integer.BYTES;
+                            long hi = bufPos + 2L * l;
+                            floatingCharSink.of(bufPos, hi);
+                            if (!Chars.utf8Decode(entity.getValue().getLo(), entity.getValue().getHi(), floatingCharSink)) {
+                                throw CairoException.instance(0).put("invalid UTF8 in value for ").put(entity.getName());
+                            }
+                            bufPos = hi;
+                        } else {
+                            try {
+                                long geohash = GeoHashes.fromString(
+                                        entity.getValue(),
+                                        0,
+                                        entity.getValue().length(),
+                                        Numbers.decodeLowShort(colTypeMeta));
+                                // as far as entity type goes, rather than store something like
+                                // NewLineProtoParser.ENTITY_TYPE_GEOHASH we simply store the storageTag
+                                // so that the writer thread can directly switch over it
+                                Unsafe.getUnsafe().putByte(bufPos, (byte) Numbers.decodeHighShort(colTypeMeta));
+                                bufPos += Byte.BYTES;
+                                Unsafe.getUnsafe().putLong(bufPos, geohash);
+                                bufPos += Long.BYTES;
+                            } catch (NumericException e) {
+                                throw CairoException.instance(0)
```

Comment:
at the point of parsing we do not have localdetails yet, we do not know whether there is a column that matches the entity's name and is declared as type geohash. All we know is that we have something that looks like a string. Later on one of the reader threads handles the string and figures out whether it indeed is to be fed to a column declared as type geohash, at which point we parse the string as a geohash and store the result. The writer thread simply gets the result, of the appropriate size and invokes the alluded `row.putGeoHash<Size>` method. The heavy lifting is done by a reader thread. The TableWriter is owned by the writer thread, no work is involved there, other than putting the result of the parse operation.

Regarding the throw, yes, well spotted, I will have a look at ignoring the offending value and instead procuring a null. Indeed otherwise it will ignore the whole line, not good.